src/linear_algebra/determinant.lean

```diff
@@ -319,4 +330,208 @@ begin
+  (e : ∀x, p x ↔ q x) : (to_square_block' M p).det = (to_square_block' M q).det :=
+by convert index_equiv_det (subtype_equiv_right e) (to_square_block' M q)
+
+/-- Let `b` map rows and columns of a square matrix `M` to `n + 1` blocks. Then
+  `upper_block_triangular_matrix M n b` says the matrix is upper block triangular. -/
+def upper_block_triangular_matrix {o : Type*} [fintype o] (M : matrix o o R) (n : ℕ)
+  (b : o → ℕ) := (∀ i, b i ≤ n) ∧ (∀ i j, b j < b i → M i j = 0)
+
+lemma upper_block_triangular_det (M : matrix m m R) (n : ℕ) (b : m → ℕ)
+  (h : upper_block_triangular_matrix M n b) :
+  M.det = ∏ k in range (n + 1), (to_square_block M b k).det :=
+begin
+  tactic.unfreeze_local_instances,
+  induction n with n hn generalizing m M b,
+  { rw [zero_add, range_one, prod_singleton],
+    have hb0 : ∀ i, b i = 0, { intro i, exact nat.le_zero_iff.mp (h.left i) },
+    convert (index_equiv_det (subtype_univ_equiv hb0) M).symm },
+  { rw prod_range_succ,
+    have h2 : (M.to_square_block' (λ (i : m), b i = n.succ)).det =
+      (M.to_square_block b n.succ).det,
+    { dunfold to_square_block, dunfold to_square_block', refl },
+    rw upper_two_block_triangular_det' M (λ i, ¬(b i = n.succ)),
+    { rw mul_comm,
+      apply congr (congr_arg has_mul.mul _),
+      { let m' := {a // ¬b a = n.succ },
+        let b' := (λ (i : m'), b ↑i),
+        have h' :
+          upper_block_triangular_matrix (M.to_square_block' (λ (i : m), ¬b i = n.succ)) n b',
+        { split,
+          { intro i, exact nat.lt_succ_iff.mp ((ne.le_iff_lt i.property).mp (h.left ↑i)) },
+          { intros i j, apply h.right ↑i ↑j }},
+        have h1 := hn (M.to_square_block' (λ (i : m), ¬b i = n.succ)) b' h',
+        rw ←fin.prod_univ_eq_prod_range,
+        rw ←fin.prod_univ_eq_prod_range at h1,
+        convert h1,
+        ext k,
+        simp only [to_square_block_def', to_square_block_def],
+        let he : {a // b' a = ↑k} ≃ {a // b a = ↑k},
+        { have hc : ∀ (i : m), (λ a, b a = ↑k) i → (λ a, ¬b a = n.succ) i,
+          { intros i hbi, rw hbi, exact ne_of_lt (fin.is_lt k) },
+          exact subtype_subtype_equiv_subtype hc },
+        apply eq.symm,
+        convert index_equiv_det he (λ (i j : {a // b a = ↑k}), M ↑i ↑j),
+        ext i j,
+        have hh : ∀ i, (he.to_fun i).val = i.val.val, { simp },
+        exact congr (congr_arg M (eq.symm (hh i))) (eq.symm (hh j)) },
+      { rw to_square_block_det M b n.succ,
+        have hh : ∀ a, ¬(λ (i : m), ¬b i = n.succ) a ↔ b a = n.succ,
+        { intro i, simp only [not_not] },
+        exact equiv_block_det M hh }},
+    { intros i hi j hj,
+      apply (h.right i), simp only [not_not] at hi,
+      rw hi,
+      exact (ne.le_iff_lt hj).mp (h.left j) }}
+end
+
+lemma upper_triangular_det (n : ℕ) (M : matrix (fin (n + 1)) (fin (n + 1)) R)
+  (h : ∀ (i j : fin (n + 1)), j < i → M i j = 0) :
+  M.det = ∏ i : (fin (n + 1)), M i i :=
```

Comment:
Isn't this true for `n` too? The determinant of a 0 x 0 matrix is 1.

Reply:
Originally I convinced myself that the n+1 version would be easier to prove and that the 0x0 case would never be needed, but you are right, doing it for n is better. The proof is ever so slightly simpler and, of course, it is nicer with n.